lib/node_modules/@stdlib/utils/parse-dsv/lib/main.js

```diff
@@ -0,0 +1,657 @@
+					'data': CRLF
+				});
+			} else {
+				this._buffer.shift();
+				tokens.push({
+					'type': TOKEN_ENUM.CR,
+					'data': CR
+				});
+			}
+			break;
+		case LF:
+			this._buffer.shift();
+			tokens.push({
+				'type': TOKEN_ENUM.LF,
+				'data': LF
+			});
+			break;
+		case this.quoteCharacter:
+			this._buffer.shift();
+			tokens.push({
+				'type': TOKEN_ENUM.QUOTE,
+				'data': this.quoteCharacter
+			});
+			break;
+		case null:
+			this._buffer.shift();
+			tokens.push({
+				'type': TOKEN_ENUM.ENDOFSTREAM,
+				'data': null
+			});
+			return tokens;
+		default:
+			this._buffer.shift();
+			tokens.push({
+				'type': TOKEN_ENUM.TEXTDATA,
+				'data': char
+			});
+		}
+	}
+	return tokens;
+}
+
+/**
+* Guesses the delimiter for input.
+* @private
+* @returns {void|Array} - the guessed delimiters if the buffer has been filled
+*/
+function guessDelimiter() {
+	var quoteMatcher;
+	var accumulator;
+	var metaCount;
+	var matches;
+	var guesses;
+	var counts;
+	var lines;
+	var input;
+	var count;
+	var rows;
+	var keys;
+	var char;
+	var row;
+	var i;
+	var j;
+	var k;
+
+	input = this._buffer.join( '' );
+	lines = input.match( this.__newlineRegex );
+	if ( lines ) {
+		lines = lines.length;
+	}
+
+	if ( !lines || lines < this.watermark ) {
+		return;
+	}
+
+	rows = input.split( this.__newlineRegex );
+
+	for ( j = 0; j < rows.length; j++ ) {
+		row = rows[ j ];
+		counts = {};
+		for ( i = 0; i < row.length; i++ ) { // count up occurences of each char in each row
+			char = row[ i ];
+			if ( !counts[ char ] && char !== this.quotechar ) {
+				matches = row.match( new RegExp( char + '(?=(?:[^"]*"[^"]*")*[^"]*$)', 'g' ) );
+				if ( matches ) {
+					counts[ char ] = matches.length;
+				} else {
+					counts[ char ] = 0;
+				}
+			}
+		}
+		rows[ j ] = counts; // this is ok to do because it won't be used again in the loops
+	}
+
+	if ( Object.keys( rows[ rows.length - 1 ] ).length < 1 ) {
+		lines -= 1;
+	}
+
+	accumulator = {};
+	for ( k = 0; k < rows.length; k++ ) {
+		keys = Object.keys( rows[ k ] );
+		for ( i = 0; i < keys.length; i++ ) {
+			char = keys[ i ];
+			count = rows[ k ][ char ];
+			metaCount = 0;
+			for ( j = 0; j < rows.length; j++ ) {
+				if ( rows[ j ][ char ] === count ) {
+					metaCount += 1;
+				}
+			}
+			if ( !accumulator[ char ] ||
+				accumulator[ char ].maxMeta < metaCount ) {
+				accumulator[ char ] = {
+					'maxMeta': metaCount, // the most common amount of times char occured count times in a row
+					'count': count // the number of times char occurred in those rows
+				};
+			}
+		}
+	}
+
+	/**
+	* Sorts entries based on their meta counts.
+	*
+	* @private
+	* @param {Array} [a] - the first entry
+	* @param {Array} [b] - the second entry
+	* @returns {Number} difference
+	*/
+	function comparator( a, b ) {
+		var difference = b[ 1 ].maxMeta - a[ 1 ].maxMeta;
+		if ( difference !== 0 ) {
+			return difference;
+		}
+
+		// prefer symbols over alphanumeric
+		char = a[ 0 ].charCodeAt( 0 );
```

Comment:
I believe `char` can be made a local variable here, and the `comparator` function can be elevated to the parent scope.